src/rustcore/src/admin.rs

```diff
@@ -0,0 +1,203 @@
+// limitations under the License.
+
+use pelikan::core::admin::AdminHandler;
+use pelikan::protocol::{PartialParseError, Protocol, QuitRequest, Serializable};
+
+use std::cell::RefCell;
+use std::io::{Result, Write};
+use std::net::SocketAddr;
+use std::rc::Rc;
+use std::time::{Duration, Instant};
+
+use tokio::net::{TcpListener, TcpStream};
+use tokio::prelude::*;
+use tokio::runtime::current_thread::spawn;
+use tokio::timer::Interval;
+
+// Uncomment once tokio updates to 0.2.0-alpha.7
+// use tokio::signal::CtrlC;
+// use futures::select;
+
+use ccommon::buf::OwnedBuf;
+use ccommon_sys::{buf, buf_sock_borrow, buf_sock_return};
+
+type Request<H> = <<H as AdminHandler>::Protocol as Protocol>::Request;
+type Response<H> = <<H as AdminHandler>::Protocol as Protocol>::Response;
+
+#[inline(never)]
+fn assert_buf_valid(buf: &mut OwnedBuf) {
+    unsafe {
+        let ptr: *mut *mut buf = std::mem::transmute(buf);
+
+        let mut value = true;
+        value = value && (**ptr).rpos as usize <= (**ptr).end as usize;
+        value = value && (**ptr).wpos as usize <= (**ptr).end as usize;
+        value = value && (**ptr).rpos as usize <= (**ptr).wpos as usize;
+
+        if !value {
+            panic!();
+        }
+    }
+}
+
+/// Process a single request stream
+async fn admin_tcp_stream_handler<H, S>(handler: Rc<RefCell<H>>, mut stream: S)
+where
+    H: AdminHandler + 'static,
+    S: AsyncWrite + AsyncRead + Unpin,
+    <H::Protocol as Protocol>::Request: QuitRequest,
+{
+    let mut sock = unsafe { buf_sock_borrow() };
+
+    let (mut rbuf, mut wbuf) = unsafe {
+        (
+            OwnedBuf::from_raw((*sock).wbuf),
+            OwnedBuf::from_raw((*sock).rbuf),
+        )
+    };
+
+    let mut tmpbuf = [0u8; 1024];
+
+    let mut req = Request::<H>::default();
+    let mut rsp = Response::<H>::default();
+
+    // let ctrlc = CtrlC::new();
+
+    'outer: loop {
+        let nbytes = match stream.read(&mut tmpbuf).await {
+            Ok(nbytes) => nbytes,
+            Err(_) => break,
+        };
+
+        // Uncomment this once tokio updates to 0.2.0-alpha.7
+        // let nbytes = select! {
+        //     res = stream.read(&mut tmpbuf).fuse() => match res {
+        //         Ok(nbytes) => nbytes,
+        //         Err(_) => break 'outer
+        //     },
+        //     _ = ctrlc => break 'outer
+        // };
+
+        let _ = rbuf.fit(rbuf.read_size() + tmpbuf.len());
+        match rbuf.write_all(&tmpbuf[..nbytes]) {
+            Ok(()) => (),
+            Err(e) => {
+                error!("Failed to expand buffer: {}", e);
+                // There really isn't anything we can do to recover the
+                // connection from this state since we just truncated
+                // part of a message. Instead, we just close the connection.
+                break;
+            }
+        };
+
+        while rbuf.read_size() > 0 {
+            if let Err(e) = req.parse(&mut rbuf) {
```

Comment:
structurally, I think it's cleaner if request handling is a separate function that is not embedded in logic that starts/ends with IO.

Reply:
I've refactored some of this out - I think that addresses the comment?